File: 0617/SeoyeonHong/거리두기 확인하기.py
from collections import deque

def solution(places):
    answer = []
    N = 5
    def check_around(r, c):
        visited = [[False for _ in range(N)] for _ in range(N)]
        q = deque([(r, c)])
        visited[r][c] = True
    
        while q:
            cr, cc = q.popleft()
            for nr, nc in [(cr+1, cc), (cr, cc+1), (cr-1, cc), (cr, cc-1)]:
                if 0 <= nr < N and 0 <= nc < N and not visited[nr][nc]: # 대기실 범위인지 확인
                    if place[nr][nc] == 'P': # 사람이 있을 경우
                        return False
                    elif place[nr][nc] == 'O': # 빈 테이블이 있을 경우
                        if abs(r - nr) + abs(c - nc) < 2:
                            q.append((nr, nc))
                            visited[nr][nc] = True
        return True
                    
    for i in range(N): # 각 대기실에 대해
        flag = True
        place = places[i]
        for r in range(N):
            for c in range(N):
                if place[r][c] == 'P': # 사람이 있을 경우
                    flag = check_around(r, c) # 맨해튼 거리 2 이하인 곳 확인
                if not flag: # 거리두기를 지키고 있지 않다면 반복문 중지
                    break
            if not flag:
                break
        answer.append(1 if flag else 0)
                                
    return answer

File: 0617/SeoyeonHong/거리두기 확인하기.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public boolean checkAround(int r, int c, int N, String[] place){
        boolean[][] visited = new boolean[N][N];
        visited[r][c] = true;
        Deque<int[]> deque = new ArrayDeque<>();
        deque.add(new int[]{r, c});
        int[] dr = {1, 0, -1, 0};
        int[] dc = {0, 1, 0, -1};

        while(!deque.isEmpty()){
            int[] cur = deque.pollFirst();
            for(int i = 0; i < 4; i++){
                int nr = cur[0] + dr[i];
                int nc = cur[1] + dc[i];

                if(0 <= nr && nr < N && 0 <= nc && nc < N && !visited[nr][nc]){
                    if(place[nr].charAt(nc) == 'P'){
                        return false;
                    }
                    else if(place[nr].charAt(nc) == 'O'){
                        if(Math.abs(r - nr) + Math.abs(c - nc) < 2){
                            deque.add(new int[] {nr, nc});
                            visited[nr][nc] = true;
                        }
                    }
                }
            }

        }

        return true;
    }

    public int[] solution(String[][] places) {
        int N = 5;
        int[] answer = new int[N];
        for(int i = 0; i < N; i++){
            String[] place = places[i];
            boolean flag = true;
            for(int r = 0; r < N; r++){
                for(int c = 0; c < N; c++){
                    if(place[r].charAt(c) == 'P'){
                        flag = checkAround(r, c, N, place);
                    }

                    if(!flag) break;
                }
                if(!flag) break;
            }

            answer[i] = flag ? 1 : 0;
        }
        return answer;
    }
}